File: src/algebra2_exponential_growth_decay_medium.py
"""
algebra2 - exponential_growth_decay (medium)
Generated: 2026-02-22T04:28:31.548601
"""

from problem_utils import *

def generate():
    problem_type = choice(['standard_growth', 'standard_decay', 'half_life', 'doubling_time', 'compound_interest', 'bacteria_growth'])
    
    if problem_type == 'standard_growth':
        # y = a * b^t where b > 1
        initial = choice([100, 200, 500, 1000, 1500, 2000])
        growth_rate = choice([1.05, 1.1, 1.15, 1.2, 1.25, 1.5, 2])
        time = randint(3, 8)
        
        ans = initial * (growth_rate ** time)
        ans = round(float(ans), 2)
        
        return problem(
            question=f"A population starts at {initial} and grows by a factor of {growth_rate} each year. What is the population after {time} years? Round to two decimal places.",
            answer=ans,
            difficulty=(1300, 1400),
            topic="algebra2/exponential_growth_decay",
            solution=steps(
                f"Use the exponential growth formula: $y = a \\cdot b^t$",
                f"Initial amount $a = {initial}$, growth factor $b = {growth_rate}$, time $t = {time}$",
                f"$y = {initial} \\cdot {growth_rate}^{{{time}}}$",
                f"$y = {initial} \\cdot {round(growth_rate**time, 4)}$",
                f"$y = {ans}$"
            ),
            calculator="scientific"
        )
    
    elif problem_type == 'standard_decay':
        # y = a * b^t where 0 < b < 1
        initial = choice([1000, 2000, 5000, 10000])
        decay_rate = choice([0.5, 0.6, 0.7, 0.75, 0.8, 0.85, 0.9])
        time = randint(4, 10)
        
        ans = initial * (decay_rate ** time)
        ans = round(float(ans), 2)
        
        return problem(
            question=f"A substance has an initial mass of {initial} grams and decays by a factor of {decay_rate} each hour. What is the mass after {time} hours? Round to two decimal places.",
            answer=ans,
            difficulty=(1300, 1400),
            topic="algebra2/exponential_growth_decay",
            solution=steps(
                f"Use the exponential decay formula: $y = a \\cdot b^t$",
                f"Initial amount $a = {initial}$, decay factor $b = {decay_rate}$, time $t = {time}$",
                f"$y = {initial} \\cdot {decay_rate}^{{{time}}}$",
                f"$y = {initial} \\cdot {round(decay_rate**time, 6)}$",
                f"$y = {ans}$"
            ),
            calculator="scientific"
        )
    
    elif problem_type == 'half_life':
        # Find amount remaining after several half-lives
        initial = choice([800, 1600, 3200, 6400])
        half_life = choice([2, 3, 4, 5])
        time = half_life * randint(2, 4)
        
        num_half_lives = time // half_life
        ans = initial * (0.5 ** num_half_lives)
        
        return problem(
            question=f"A radioactive substance has a half-life of {half_life} years. If you start with {initial} grams, how many grams remain after {time} years?",
            answer=ans,
            difficulty=(1400, 1500),
            topic="algebra2/exponential_growth_decay",
            solution=steps(
                f"Number of half-lives = $\\frac{{{time}}}{{{half_life}}} = {num_half_lives}$",
                f"After each half-life, half of the substance remains",
                f"Amount remaining = ${initial} \\cdot \\left(\\frac{{1}}{{2}}\\right)^{{{num_half_lives}}}$",
                f"Amount remaining = ${initial} \\cdot {0.5**num_half_lives}$",
                f"Amount remaining = ${ans}$ grams"
            ),
            calculator="scientific"
        )
    
    elif problem_type == 'doubling_time':
        # Find amount after several doubling periods
        initial = choice([50, 100, 200, 400])
        doubling_time = choice([2, 3, 4, 6])
        time = doubling_time * randint(2, 4)
        
        num_doublings = time // doubling_time
        ans = initial * (2 ** num_doublings)
        
        return problem(
            question=f"A bacteria culture doubles every {doubling_time} hours. If it starts with {initial} bacteria, how many bacteria are there after {time} hours?",
            answer=ans,
            difficulty=(1400, 1500),
            topic="algebra2/exponential_growth_decay",
            solution=steps(
                f"Number of doubling periods = $\\frac{{{time}}}{{{doubling_time}}} = {num_doublings}$",
                f"After each doubling period, the population doubles",
                f"Population = ${initial} \\cdot 2^{{{num_doublings}}}$",
                f"Population = ${initial} \\cdot {2**num_doublings}$",
                f"Population = ${ans}$ bacteria"
            ),
            calculator="scientific"
        )
    
    elif problem_type == 'compound_interest':
        # A = P(1 + r)^t
        principal = choice([1000, 2000, 5000, 10000])
        rate_percent = choice([3, 4, 5, 6, 8])
        rate = rate_percent / 100
        time = randint(5, 15)
        
        ans = principal * ((1 + rate) ** time)
        ans = round(float(ans), 2)
        
        return problem(
            question=f"An investment of $\\${principal}$ earns ${rate_percent}\\%$ interest compounded annually. What is the value of the investment after ${time}$ years? Round to two decimal places.",
            answer=ans,
            difficulty=(1400, 1600),
            topic="algebra2/exponential_growth_decay",
            solution=steps(
                f"Use compound interest formula: $A = P(1 + r)^t$",
                f"Principal $P = {principal}$, rate $r = {rate}$, time $t = {time}$",
                f"$A = {principal}(1 + {rate})^{{{time}}}$",
                f"$A = {principal} \\cdot {1 + rate}^{{{time}}}$",
                f"$A = {principal} \\cdot {round((1 + rate)**time, 6)}$",
                f"$A = \\${ans}$"
            ),
            calculator="scientific"
        )
    
    else:  # bacteria_growth with percentage
        initial = choice([100, 250, 500, 1000])
        growth_percent = choice([15, 20, 25, 30, 40, 50])
        growth_rate = 1 + (growth_percent / 100)
        time = randint(3, 6)
        
        ans = initial * (growth_rate ** time)
        ans = round(float(ans), 2)
        
        return problem(
            question=f"A bacteria population starts at {initial} and increases by ${growth_percent}\\%$ each day. What is the population after {time} days? Round to two decimal places.",
            answer=ans,
            difficulty=(1500, 1600),
            topic="algebra2/exponential_growth_decay",
            solution=steps(
                f"Growth rate as decimal: ${growth_percent}\\% = {growth_percent/100}$",
                f"Growth factor: $1 + {growth_percent/100} = {growth_rate}$",
                f"Use formula: $y = {initial} \\cdot {growth_rate}^t$",
                f"$y = {initial} \\cdot {growth_rate}^{{{time}}}$",
                f"$y = {initial} \\cdot {round(growth_rate**time, 6)}$",
                f"$y = {ans}$"
            ),
            calculator="scientific"
        )

emit(generate())

File: src/algebra2_exponential_growth_decay_medium.jl
# algebra2 - exponential_growth_decay (medium)
# Generated: 2026-03-08T20:18:53.563798

include(joinpath(@__DIR__, "..", "julia", "src", "ProblemUtils.jl"))
using .ProblemUtils

@script x t begin
    set_topic!("algebra2/exponential_growth_decay")
    
    problem_type = rand(1:5)
    
    if problem_type == 1
        # Exponential growth: population/bacteria/investment
        context = choice(["bacteria", "population", "investment"])
        
        if context == "bacteria"
            initial = randint(100, 2000)
            growth_rate = choice([0.05, 0.08, 0.10, 0.12, 0.15, 0.20, 0.25])
            time_val = randint(3, 12)
            
            final_amount = round(Int, initial * (1 + growth_rate)^time_val)
            
            problem(
                question="A bacteria culture starts with $(initial) bacteria and grows at a rate of $(round(Int, growth_rate*100))% per hour. How many bacteria will there be after $(time_val) hours? Round to the nearest whole number.",
                answer=final_amount,
                difficulty=(1200, 1400),
                solution=steps(
                    "Use exponential growth formula: \$N(t) = N_0(1 + r)^t\$",
                    sol("Substitute values", "N($(time_val)) = $(initial)(1 + $(growth_rate))^{$(time_val)}"),
                    sol("Calculate", "N($(time_val)) = $(initial)($(1 + growth_rate))^{$(time_val)} = $(final_amount)"),
                    "After $(time_val) hours, there are $(final_amount) bacteria"
                ),
                time=120
            )
            
        elseif context == "population"
            initial = randint(5000, 50000)
            growth_rate = choice([0.02, 0.03, 0.04, 0.05, 0.06])
            time_val = randint(5, 20)
            
            final_amount = round(Int, initial * (1 + growth_rate)^time_val)
            
            problem(
                question="A city has a population of $(initial) and grows at $(round(Int, growth_rate*100))% per year. What will the population be after $(time_val) years? Round to the nearest whole number.",
                answer=final_amount,
                difficulty=(1200, 1400),
                solution=steps(
                    "Use exponential growth formula: \$P(t) = P_0(1 + r)^t\$",
                    sol("Substitute", "P($(time_val)) = $(initial)(1.0$(lpad(round(Int, growth_rate*100), 2, '0')))^{$(time_val)}"),
                    sol("Calculate", "P($(time_val)) = $(final_amount)"),
                    "The population after $(time_val) years is $(final_amount)"
                ),
                time=120
            )
            
        else  # investment
            initial = randint(1000, 10000)
            rate = choice([0.03, 0.04, 0.05, 0.06, 0.07, 0.08])
            time_val = randint(5, 15)
            
            final_amount = round(initial * (1 + rate)^time_val, digits=2)
            
            problem(
                question="An investment of \\\$$(initial) grows at $(round(Int, rate*100))% per year, compounded annually. What is the value after $(time_val) years? Round to the nearest cent.",
                answer=final_amount,
                difficulty=(1200, 1400),
                solution=steps(
                    "Use compound interest formula: \$A = P(1 + r)^t\$",
                    sol("Substitute", "A = $(initial)(1 + $(rate))^{$(time_val)}"),
                    sol("Calculate", "A = $(initial)($(1 + rate))^{$(time_val)} \\approx $(final_amount)"),
                    "The investment is worth \\\$$(final_amount) after $(time_val) years"
                ),
                time=120
            )
        end
        
    elseif problem_type == 2
        # Exponential decay: radioactive/depreciation/medicine
        context = choice(["radioactive", "car", "medicine"])
        
        if context == "radioactive"
            initial = randint(100, 500)
            half_life = choice([2, 3, 4, 5, 6, 8, 10])
            time_val = randint(2, 5) * half_life
            
            remaining = initial * (0.5)^(time_val / half_life)
            
            problem(
                question="A radioactive substance has a half-life of $(half_life) years. If you start with $(initial) grams, how much remains after $(time_val) years?",
                answer=remaining,
                difficulty=(1300, 1500),
                solution=steps(
                    "Use half-life formula: \$N(t) = N_0 \\left(\\frac{1}{2}\\right)^{t/h}\$",
                    sol("Substitute", "N($(time_val)) = $(initial)\\left(\\frac{1}{2}\\right)^{$(time_val)/$(half_life)}"),
                    sol("Calculate", "N($(time_val)) = $(initial)\\left(\\frac{1}{2}\\right)^{$(div(time_val, half_life))} = $(remaining)"),
                    "After $(time_val) years, $(remaining) grams remain"
                ),
                time=150
            )
            
        elseif context == "car"
            initial = randint(15000, 40000)
            decay_rate = choice([0.12, 0.15, 0.18, 0.20, 0.25])
            time_val = randint(3, 8)
            
            final_value = round(initial * (1 - decay_rate)^time_val, digits=2)
            
            problem(
                question="A car purchased for \\\$$(initial) depreciates at $(round(Int, decay_rate*100))% per year. What is its value after $(time_val) years? Round to the nearest cent.",
                answer=final_value,
                difficulty=(1300, 1500),
                solution=steps(
                    "Use exponential decay formula: \$V(t) = V_0(1 - r)^t\$",
                    sol("Substitute", "V($(time_val)) = $(initial)(1 - $(decay_rate))^{$(time_val)}"),
                    sol("Calculate", "V($(time_val)) = $(initial)($(1 - decay_rate))^{$(time_val)} \\approx $(final_value)"),
                    "The car's value after $(time_val) years is \\\$$(final_value)"
                ),
                time=150
            )
            
        else  # medicine
            initial = randint(100, 800)
            half_life = choice([2, 3, 4, 6, 8])
            time_val = randint(1, 3) * half_life
            
            remaining = initial * (0.5)^(time_val / half_life)
            
            problem(
                question="A medication has a half-life of $(half_life) hours in the body. If $(initial) mg is administered, how much remains after $(time_val) hours?",
                answer=remaining,
                difficulty=(1300, 1500),
                solution=steps(
                    "Use half-life decay: \$A(t) = A_0 \\left(\\frac{1}{2}\\right)^{t/h}\$",
                    sol("Substitute", "A($(time_val)) = $(initial)\\left(\\frac{1}{2}\\right)^{$(time_val)/$(half_life)}"),
                    sol("Calculate", "A($(time_val)) = $(initial) \\cdot \\frac{1}{$(2^div(time_val, half_life))} = $(remaining)"),
                    "After $(time_val) hours, $(remaining) mg remain in the body"
                ),
                time=150
            )
        end
        
    elseif problem_type == 3
        # Solve for time given initial and final amounts
        initial = randint(1000, 5000)
        rate = choice([0.04, 0.05, 0.06, 0.08, 0.10])
        time_val = randint(5, 15)
        final_amount = round(Int, initial * (1 + rate)^time_val)
        
        problem(
            question="An investment of \\\$$(initial) grows at $(round(Int, rate*100))% per year. How many years will it take to reach \\\$$(final_amount)? Round to the nearest year.",
            answer=time_val,
            difficulty=(1500, 1700),
            solution=steps(
                sol("Set up equation", "$(final_amount) = $(initial)(1 + $(rate))^t"),
                sol("Divide both sides", "\\frac{$(final_amount)}{$(initial)} = ($(1 + rate))^t"),
                sol("Take logarithm", "t = \\frac{\\log($(final_amount/initial))}{\\log($(1 + rate))} \\approx $(time_val)"),
                "It takes approximately $(time_val) years"
            ),
            time=180
        )
        
    elseif problem_type == 4
        # Continuous growth: A = Pe^(rt)
        initial = randint(1000, 8000)
        rate = choice([0.03, 0.04, 0.05, 0.06, 0.07])
        time_val = randint(5, 15)
        
        e_val = 2.71828
        final_amount = round(initial * e_val^(rate * time_val), digits=2)
        
        problem(
            question="An investment of \\\$$(initial) grows continuously at $(round(Int, rate*100))% per year. What is the value after $(time_val) years? Use \$e \\approx 2.71828\$ and round to the nearest cent.",
            answer=final_amount,
            difficulty=(1500, 1700),
            solution=steps(
                "Use continuous growth formula: \$A = Pe^{rt}\$",
                sol("Substitute", "A = $(initial)e^{$(rate) \\cdot $(time_val)}"),
                sol("Calculate exponent", "A = $(initial)e^{$(rate * time_val)} \\approx $(initial)($(round(e_val^(rate * time_val), digits=5)))"),
                sol("Final answer", "A \\approx $(final_amount)"),
                "The investment is worth \\\$$(final_amount)"
            ),
            time=180
        )
        
    else  # problem_type == 5
        # Word problem: find growth/decay rate
        initial = randint(2000, 10000)
        time_val = randint(5, 12)
        multiplier = choice([2, 3, 4])
        final_amount = initial * multiplier
        
        # Solve for r: final = initial(1 + r)^t
        # multiplier = (1 + r)^t
        # r = multiplier^(1/t) - 1
        rate = round((multiplier^(1/time_val) - 1) * 100, digits=2)
        
        problem(
            question="A population grows from $(initial) to $(final_amount) in $(time_val) years. What is the annual growth rate as a percentage? Round to two decimal places.",
            answer=rate,
            difficulty=(1600, 1800),
            solution=steps(
                sol("Set up equation", "$(final_amount) = $(initial)(1 + r)^{$(time_val)}"),
                sol("Divide both sides", "$(multiplier) = (1 + r)^{$(time_val)}"),
                sol("Take $(time_val)-th root", "1 + r = $(multiplier)^{1/$(time_val)} \\approx $(round(multiplier^(1/time_val), digits=5))"),
                sol("Solve for r", "r \\approx $(round(multiplier^(1/time_val) - 1, digits=5)) = $(rate)\\%"),
                "The annual growth rate is approximately $(rate)%"
            ),
            time=200
        )
    end
end